File: client/src/main/java/net/treset/minecraftlauncher/sync/AllSynchronizer.java
package net.treset.minecraftlauncher.sync;

import javafx.util.Pair;
import net.treset.mc_version_loader.launcher.LauncherManifest;
import net.treset.minecraftlauncher.data.InstanceData;
import net.treset.minecraftlauncher.data.LauncherFiles;
import net.treset.minecraftlauncher.util.SyncUtil;
import net.treset.minecraftlauncher.util.exception.FileLoadException;

import java.io.IOException;
import java.util.ArrayList;

public class AllSynchronizer extends FileSynchronizer {
    private final LauncherFiles files;

    public AllSynchronizer(LauncherFiles files, SyncCallback callback) {
        super(callback);
        this.files = files;
    }

    @Override
    public void upload() throws IOException {
        synchronize(true);
    }

    @Override
    public void download() throws IOException {
        synchronize(false);
    }

    private void synchronize(boolean upload) throws IOException {
        ArrayList<IOException> exceptions = new ArrayList<>();
        files.getInstanceComponents().parallelStream().forEach((details) -> {
            if(!SyncUtil.isSyncing(details.getKey())) {
                return;
            }
            InstanceData data;
            try {
                data = InstanceData.of(details, files);
            } catch (FileLoadException e) {
                exceptions.add(new IOException(e));
                return;
            }
            InstanceSynchronizer synchronizer = new InstanceSynchronizer(data, files, callback);
            try {
                if(upload) {
                    synchronizer.upload();
                } else {
                    synchronizer.download();
                }
            } catch (IOException e) {
                exceptions.add(e);
            }
        });
        ArrayList<LauncherManifest> manifests = new ArrayList<>();
        manifests.addAll(files.getSavesComponents());
        manifests.addAll(files.getResourcepackComponents());
        manifests.addAll(files.getOptionsComponents());
        manifests.addAll(files.getModsComponents().stream().map(Pair::getKey).toList());
        manifests.parallelStream().forEach((manifest) -> {
            if(!SyncUtil.isSyncing(manifest)) {
                return;
            }
            ManifestSynchronizer synchronizer = new ManifestSynchronizer(manifest, files, callback);
            try {
                if(upload) {
                    synchronizer.upload();
                } else {
                    synchronizer.download();
                }
            } catch (IOException e) {
                exceptions.add(e);
            }
        });
        if(!exceptions.isEmpty()) {
            throw exceptions.get(0);
        }
    }
}

File: client/composeApp/src/commonMain/kotlin/net/treset/treelauncher/backend/sync/AllSynchronizer.kt
package net.treset.treelauncher.backend.sync

import net.treset.treelauncher.backend.data.InstanceData
import net.treset.treelauncher.backend.data.LauncherFiles
import net.treset.treelauncher.backend.data.manifest.ComponentManifest
import net.treset.treelauncher.backend.util.exception.FileLoadException
import java.io.IOException

class AllSynchronizer(private val files: LauncherFiles, callback: SyncCallback?) : FileSynchronizer(callback) {

    @Throws(IOException::class)
    override fun upload() {
        synchronize(true)
    }

    @Throws(IOException::class)
    override fun download() {
        synchronize(false)
    }

    @Throws(IOException::class)
    private fun synchronize(upload: Boolean) {
        val exceptions: MutableList<IOException> = mutableListOf()
        files.instanceComponents.forEach { details ->
            if (!SyncService.isSyncing(details.first)) {
                return@forEach
            }
            val data = try {
                InstanceData.of(details, files)
            } catch (e: FileLoadException) {
                exceptions.add(IOException(e))
                return@forEach
            }
            val synchronizer = InstanceSynchronizer(data, files, callback)
            try {
                if (upload) {
                    synchronizer.upload()
                } else {
                    synchronizer.download()
                }
            } catch (e: IOException) {
                exceptions.add(e)
            }
        }
        val manifests: MutableList<ComponentManifest> = mutableListOf()
        manifests.addAll(files.savesComponents)
        manifests.addAll(files.resourcepackComponents)
        manifests.addAll(files.optionsComponents)
        manifests.addAll(files.modsComponents.map{it.first}.toList())
        manifests.parallelStream().forEach { manifest: ComponentManifest ->
            if (!SyncService.isSyncing(manifest)) {
                return@forEach
            }
            val synchronizer = ManifestSynchronizer(manifest, files, callback)
            try {
                if (upload) {
                    synchronizer.upload()
                } else {
                    synchronizer.download()
                }
            } catch (e: IOException) {
                exceptions.add(e)
            }
        }
        if (exceptions.isNotEmpty()) {
            throw exceptions[0]
        }
    }
}
